tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py

```diff
@@ -0,0 +1,146 @@
+"""
+
+import logging
+import os
+import random
+import subprocess
+import sys
+import time
+
+from boardfarm.devices import debian
+from environment import ALEntityDocker, _get_bridge_interface
+
+rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..'))
+
+logging.basicConfig(stream=sys.stdout, format='%(message)s')
+logger = logging.getLogger('bft')
+logger.setLevel(logging.INFO)  # DEBUG, INFO, WARNING, ERROR, CRITICAL
+
+
+class PrplMeshDocker(debian.DebianBox):
+    '''
+    Dockerized prplMesh device
+    '''
+    model = ("prplmesh_docker")
+
+    def __init__(self, *args, **kwargs):
+        self.args = args
+        self.kwargs = kwargs
+
+        config = kwargs["config"] if kwargs.get("config", None) else kwargs
+
+        # String buffer which stores all calls to device's consoles
+        self.log_calls = ""
+
+        # String buffer with device's console log
+        self.log = ""
+
+        # List of device's consoles test can interact with
+        self.consoles = [self]
+
+        # Getting unic ID to distinguish devices and network they belong to
+        self.unique_id = os.getenv("SUDO_USER", os.getenv("USER", ""))
+
+        self.name = "-".join((config.get("name", "prplmesh_docker"), self.unique_id))
+        self.role = config.get("role", "agent")
+        self.cleanup_cmd = config.get("cleanup_cmd", None)
+        self.conn_cmd = config.get("conn_cmd", None)
+        self.mac = config.get("mac", self._get_rand_mac(oct_amount=4))
+        self.use_tcpdump = config.get("use_tcpdump", False)
+        self.docker_network = config.get("docker_network",
+                                         "prplMesh-net-{}".format(self.unique_id))
+
+        if self.cleanup_cmd:
+            self._run_shell_cmd(self.cleanup_cmd)
+
+        docker_run_cmd = os.path.join(rootdir, "tools", "docker", "run.sh") \
+            + " --verbose --detach --force --name {}".format(self.name) \
+            + " --mac {}".format(self.mac) \
+            + " --network {}".format(self.docker_network)
+
+        if self.role == "both":
+            # Spawn dockerized controller + agent in single device
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
+        elif self.role == "controller":
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
+        else:
+            # Spawn dockerized agent
+            self._run_shell_cmd(docker_run_cmd + " start-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=False)
+
+        self.gw_iface = _get_bridge_interface(self.docker_network)
+
+        self.check_status()
```

Comment:
It would be more appropriate to have this as a test instead of doing it in the constructor.

And that should be a separate commit.

Reply:
> It would be more appropriate to have this as a test instead of doing it in the constructor.
> 
> And that should be a separate commit.

That method is a "health check" and must be present in the device class:
it is used by boardfarm to indicate that spawned device instance is ready for test and also it is used after test to insure that device still operational.
So creating a testcase from that method will be wrong, as we have to run it after each test in boardfarm testsuite - that's a mess IMO.

Reply:
Actually, AFAICS in the source code of boardfarm, this method is only called after a failed test, as a way to distinguish between a failing test and a failing device.

But regardless, your point still stands: it doesn't make much sense to implement this as a separate test. And it *does* make sense to check the status in the constructor, since indeed boardfarm assumes that the device is fully operational when the constructor finishes.